python/tests/test_conversions.py

```diff
@@ -0,0 +1,422 @@
+        (FixedType(3), "foo", bytearray(b"foo")),
+        (BinaryType(), "foo", b"foo"),
+        (None, None, None),
+    ],
+)
+def test_from_partition_value_to_py(primitive_type, partition_value_as_str, expected_result):
+    """Test converting a partition value to a python built-in"""
+    assert conversions.from_partition_value_to_py(primitive_type, partition_value_as_str) == expected_result
+
+
+@pytest.mark.parametrize(
+    "primitive_type, b, result",
+    [
+        (BooleanType(), b"\x00", False),
+        (BooleanType(), b"\x01", True),
+        (IntegerType(), b"\xd2\x04\x00\x00", 1234),
+        (LongType(), b"\xd2\x04\x00\x00\x00\x00\x00\x00", 1234),
+        (DoubleType(), b"\x8d\x97\x6e\x12\x83\xc0\xf3\x3f", 1.2345),
+        (DateType(), bytes([232, 3, 0, 0]), 1000),
+        (DateType(), b"\xd2\x04\x00\x00", 1234),
+        (TimeType(), bytes([16, 39, 0, 0, 0, 0, 0, 0]), 10000),
+        (TimeType(), b"\x00\xe8vH\x17\x00\x00\x00", 100000000000),
+        (TimestamptzType(), bytes([128, 26, 6, 0, 0, 0, 0, 0]), 400000),
+        (TimestamptzType(), b"\x00\xe8vH\x17\x00\x00\x00", 100000000000),
+        (TimestampType(), bytes([128, 26, 6, 0, 0, 0, 0, 0]), 400000),
+        (TimestampType(), b"\x00\xe8vH\x17\x00\x00\x00", 100000000000),
+        (StringType(), bytes([65, 66, 67]), "ABC"),
+        (StringType(), b"foo", "foo"),
+        (
+            UUIDType(),
+            bytes([247, 156, 62, 9, 103, 124, 75, 189, 164, 121, 63, 52, 156, 183, 133, 231]),
+            uuid.UUID("f79c3e09-677c-4bbd-a479-3f349cb785e7"),
+        ),
+        (UUIDType(), b"\xf7\x9c>\tg|K\xbd\xa4y?4\x9c\xb7\x85\xe7", uuid.UUID("f79c3e09-677c-4bbd-a479-3f349cb785e7")),
+        (FixedType(3), bytes([97, 98]), bytes("ab", "utf8")),
+        (FixedType(3), b"foo", b"foo"),
+        (BinaryType(), bytearray("Z", "utf8"), bytes([90])),
+        (BinaryType(), b"foo", b"foo"),
+        (DecimalType(5, 2), b"\x30\x39", Decimal("123.45")),
+        (DecimalType(7, 4), b"\x00\x12\xd6\x87", Decimal("123.4567")),
+        (DecimalType(7, 4), b"\xff\xed\x29\x79", Decimal("-123.4567")),
+    ],
+)
+def test_from_bytes(primitive_type, b, result):
+    """Test converting from bytes"""
+    assert conversions.from_bytes(primitive_type, b) == result
+
+
+@pytest.mark.parametrize(
+    "primitive_type, b, approximate_result, approximation",
+    [
+        (FloatType(), b"\x19\x04\x9e?", 1.2345, 5),
+    ],
+)
+def test_from_bytes_approximately(primitive_type, b, approximate_result, approximation):
+    """Test approximate equality when converting from bytes
+
+    Note: This tests approximate equality because floating point numbers in python are implemented
+    using a double in C. Therefore a 32-bit (single precision) float is unpacked to 64-bit (double precision)
+    float in python which introduces some imprecision.
+    """
+    assert conversions.from_bytes(primitive_type, b) == pytest.approx(approximate_result, approximation)
```

Comment:
I don't think this should be an approximate result. Instead, can you send the expected result through pack/unpack to produce the same 32-bit value?

Reply:
Ah right, added the pack/unpack and it works without any approximation.